ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala

```diff
@@ -0,0 +1,309 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.cache
+
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicLong
+
+import akka.NotUsed
+import akka.stream.scaladsl.{Flow, Sink}
+import com.daml.ledger.participant.state.index.v2.ContractStore
+import com.daml.ledger.participant.state.v1.Offset
+import com.daml.lf.transaction.GlobalKey
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.metrics.{Metrics, Timed}
+import com.daml.platform.store.cache.ContractStateValue._
+import com.daml.platform.store.cache.ContractKeyStateValue._
+import com.daml.platform.store.cache.MutableCacheBackedContractStore.{
+  ContractNotFound,
+  EmptyContractIds,
+}
+import com.daml.platform.store.dao.events.ContractStateEvent
+import com.daml.platform.store.dao.events.ContractStateEvent.LedgerEndMarker
+import com.daml.platform.store.interfaces.LedgerDaoContractsReader
+import com.daml.platform.store.interfaces.LedgerDaoContractsReader.{
+  ActiveContract,
+  ArchivedContract,
+  ContractState,
+  KeyState,
+}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class MutableCacheBackedContractStore(
+    metrics: Metrics,
+    contractsReader: LedgerDaoContractsReader,
+    signalNewLedgerHead: Offset => Unit,
+    private[cache] val keyCache: StateCache[GlobalKey, ContractKeyStateValue],
+    private[cache] val contractsCache: StateCache[ContractId, ContractStateValue],
+)(implicit
+    executionContext: ExecutionContext
+) extends ContractStore {
+  private val logger = ContextualizedLogger.get(getClass)
+  private[cache] val cacheOffset = new AtomicLong(0L)
+
+  def consumeFrom(implicit
+      loggingContext: LoggingContext
+  ): Flow[ContractStateEvent, Unit, NotUsed] =
+    Flow[ContractStateEvent]
+      .wireTap(debugEvents(_))
+      .alsoTo(Sink.foreach(updateCaches))
+      .map(updateOffsets)
+
+  override def lookupActiveContract(readers: Set[Party], contractId: ContractId)(implicit
+      loggingContext: LoggingContext
+  ): Future[Option[Contract]] =
+    contractsCache
+      .get(contractId)
+      .map(contractStateToResponse(_, readers, contractId))
```

Comment:
If you moved this mapping to after the `getOrElse` then you should be able to avoid passing in the `readers` and duplicating the call to this transformation.

Reply:
Nice eye, much better now